File: csharp/Decorator.cs
using System;

class Coffee {
    public virtual int Cost() => 5;
}

class CoffeeDecorator : Coffee {
    protected Coffee _coffee;

    public CoffeeDecorator(Coffee coffee) {
        _coffee = coffee;
    }

    public override int Cost() => _coffee.Cost();
}

class MilkDecorator : CoffeeDecorator {
    public MilkDecorator(Coffee coffee) : base(coffee) { }

    public override int Cost() => base.Cost() + 2;
}

class SugarDecorator : CoffeeDecorator {
    public SugarDecorator(Coffee coffee) : base(coffee) { }

    public override int Cost() => base.Cost() + 1;
}

class Program {
    static void Main() {
        var coffee = new Coffee();
        var milk = new MilkDecorator(coffee);
        var sugar = new SugarDecorator(coffee);
        var both = new SugarDecorator(milk);

        Console.WriteLine(coffee.Cost());      // 5
        Console.WriteLine(milk.Cost());        // 7
        Console.WriteLine(sugar.Cost());       // 6
        Console.WriteLine(both.Cost());        // 8
    }
}

File: cpp/decorator.cpp
#include <iostream>
using namespace std;

class Coffee {
public:
    virtual int cost() { return 5; }
    virtual ~Coffee() = default;
};

class CoffeeDecorator : public Coffee {
protected:
    Coffee* coffee;

public:
    CoffeeDecorator(Coffee* c) : coffee(c) {}
    int cost() override { return coffee->cost(); }
};

class MilkDecorator : public CoffeeDecorator {
public:
    MilkDecorator(Coffee* c) : CoffeeDecorator(c) {}
    int cost() override { return coffee->cost() + 2; }
};

class SugarDecorator : public CoffeeDecorator {
public:
    SugarDecorator(Coffee* c) : CoffeeDecorator(c) {}
    int cost() override { return coffee->cost() + 1; }
};

int main() {
    Coffee base;
    MilkDecorator milk(&base);
    SugarDecorator sugar(&base);
    SugarDecorator both(&milk);

    cout << base.cost() << endl;         // 5
    cout << milk.cost() << endl;         // 7
    cout << sugar.cost() << endl;        // 6
    cout << both.cost() << endl;         // 8

    return 0;
}